app/src/main/java/com/example/drawingApp/MainActivity.kt

```diff
@@ -1,23 +1,118 @@
+import android.widget.TextView
+import android.widget.Toast
 import androidx.activity.viewModels
+import androidx.appcompat.app.AlertDialog
 import androidx.appcompat.app.AppCompatActivity
+import androidx.core.view.isVisible
+import androidx.core.widget.addTextChangedListener
 
 class MainActivity : AppCompatActivity() {
-    private val model: MainViewModel by viewModels()
+    private val model: DrawingViewModel by viewModels()
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
-
+        enableStrictMode()
         val colorCircleView = findViewById<SelectedColorView>(R.id.colorCircle)
         val drawingFieldView = findViewById<DrawingFieldView>(R.id.drawField)
 
+        drawingFieldView.onLineDrawing = {
+            model.activeBitmap = it
+        }
+
+        findViewById<TextView>(R.id.newLayer).setOnClickListener {
+            showLayerAlertDialog()
+        }
+
+        model.activeBitmap?.let {
+            drawingFieldView.setBitmap(it)
+            drawingFieldView.invalidate()
+        }
+
         colorCircleView.onColorChange = { stateColor ->
             model.primaryColor = stateColor
             drawingFieldView.updateColor(stateColor)
         }
 
+        if (model.isLayerCreationDialogOpen) showLayerAlertDialog()
+
+    }
+
+    fun showLayerAlertDialog() {
+        val inflater = LayoutInflater.from(this@MainActivity)
+        val view = inflater.inflate(R.layout.layer_dialog, null)
+        val layerField = view.findViewById<EditText>(R.id.newLayerField)
+        val errorText = view.findViewById<TextView>(R.id.dialogCharCountError)
+        var layerCounter = 1
+
+        //find the first available layer name not taken
+        while (model.layerNames.contains(getString(R.string.new_layer_hint, layerCounter))) {
+            layerCounter += 1
+        }
+        layerField.setText(getString(R.string.new_layer_hint, layerCounter))
+
+        //needs to be after the set text or else the textchanged listener will overwrite the previous text needed for rotations
+        layerField.addTextChangedListener {
+            model.currentLayerCreationDialogText = layerField.text.toString()
+            errorText.isVisible = it.toString().trim().length > MAX_CHAR_LEN
+            if (errorText.isVisible) {
+                layerField.setText(layerField.text.toString().dropLast(1))
+                //visibility needs to be after or else the dropLast will make it never visible
+                errorText.visibility = View.VISIBLE
+            }
+        }
+
+        val dialog = AlertDialog.Builder(this).apply {
+            setView(view)
+        }.create()
+
+        view.findViewById<TextView>(R.id.newLayerCancel).setOnClickListener {
+            dialog.dismiss()
+        }
+        view.findViewById<TextView>(R.id.newLayerOk).setOnClickListener {
+            errorText.visibility = View.INVISIBLE
```

Comment:
Shouldn't this check for duplicates? And in that case, why hide the errorText?

Reply:
it can do that, it wasn't specified so I hadn't done that, but I will implement a check.
The reason the error text was being hidden was because the text for the layer could only conform to 16 characters which means that it would've been valid in all cases without checking for a duplicate layer with the same name.

Will update with a check and slightly different logic.